core/src/impl/Kokkos_Profiling.hpp

```diff
@@ -129,6 +130,149 @@ void resume_tools();
 EventSet get_callbacks();
 void set_callbacks(EventSet new_events);
 }  // namespace Experimental
+
+namespace Experimental {
+
+// forward declarations
+size_t get_new_context_id();
+size_t get_current_context_id();
+
+namespace Impl {
+
+static std::map<std::string, Kokkos::Tools::Experimental::TeamSizeTuner>
+    team_tuners;
+
+template <class ExecPolicy, class Functor, typename TagType>
+void tune_policy(const size_t tuning_context, const std::string& label,
+                 ExecPolicy& policy, const Functor& functor,
+                 const TagType& tag){};
+
+template <class Functor, class TagType, class... Properties>
+void tune_policy(const size_t tuning_context, const std::string& label,
+                 Kokkos::TeamPolicy<Properties...>& policy,
+                 const Functor& functor, const TagType& tag) {
+  if (policy.auto_team_size() || policy.auto_vector_length()) {
+    if (team_tuners.find(label) == team_tuners.end()) {
+      team_tuners.emplace(label, Kokkos::Tools::Experimental::TeamSizeTuner(
+                                     label, policy, functor, tag));
+    }
+    auto& tuner = team_tuners[label];
+    tuner.tune(policy, tuning_context);
+  }
+}
+
+template <class ExecPolicy, class Functor, typename TagType>
+void report_policy_results(const size_t tuning_context,
+                           const std::string& label, ExecPolicy& policy,
+                           const Functor& functor, const TagType& tag){};
+
+template <class Functor, class TagType, class... Properties>
+void report_policy_results(const size_t tuning_context,
+                           const std::string& label,
+                           Kokkos::TeamPolicy<Properties...> policy,
+                           const Functor& functor, const TagType& tag) {
+  if (policy.auto_team_size() || policy.auto_vector_length()) {
+    auto& tuner = team_tuners[label];
+    tuner.end(tuning_context);
+  }
+}
+}  // namespace Impl
+
+template <class ExecPolicy, class FunctorType>
+void begin_parallel_for(ExecPolicy& policy, FunctorType& functor,
+                        const std::string& label, uint64_t& kpID) {
+  if (Kokkos::Tools::profileLibraryLoaded()) {
+    Kokkos::Impl::ParallelConstructName<FunctorType,
+                                        typename ExecPolicy::work_tag>
+        name(label);
+    Kokkos::Tools::beginParallelFor(
+        name.get(), Kokkos::Profiling::Experimental::device_id(policy.space()),
+        &kpID);
+  }
+#ifdef KOKKOS_ENABLE_TUNING
+  size_t context_id = Kokkos::Tools::Experimental::get_new_context_id();
```

Comment:
Are we getting a new context ID every time we execute a parallel_for? Doesn't this mean that the "same" parallel will be a different context every time? How do we 'tune' if we can't connect parallel_for's to each other?

Reply:
I might just need to redo #2422, since everybody seems to think that a context is some permanent thing. A context is a just a way of saying when values go in and out of scope. You know to connect parallel_x's to each other if you are providing the same output types to the same set of input types